File: taipy/gui/utils/patch.py
# Copyright 2021-2025 Avaiga Private Limited
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
# the License. You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
# an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
# specific language governing permissions and limitations under the License.

import typing as t

from . import _MapDict


def _patch_value(value: t.Any, change: t.Optional[dict] = None, remove: t.Optional[dict] = None) -> t.Any:  # noqa: C901
    original_value = value
    if isinstance(value, _MapDict):
        value = value._dict
    # TODO handle dataframe
    if isinstance(value, dict):
        if change:
            for k, v in change.items():
                if (
                    isinstance(value.get(k, None), (dict, list, _MapDict))
                    and isinstance(v, dict)
                ):
                    _patch_value(value[k], v)
                else:
                    value[k] = v
        if remove:
            for k, v in remove.items():
                if k in value:
                    if isinstance(v, dict):
                        _patch_value(value[k], remove=v)
                    else:
                        del value[k]
    elif isinstance(value, list):
        if change:
            for k, v in change.items():
                if isinstance(k, int):
                    insert = k < 0
                    idx = k if k >= 0 else -1 - k
                    if idx < len(value):
                        if isinstance(v, dict):
                            _patch_value(value[idx], v)
                        if isinstance(v, list):
                            if insert:
                                value[idx:idx] = v
                            else:
                                for jdx, nv in enumerate(v, start=idx):
                                    if jdx < len(value):
                                        if isinstance(nv, dict):
                                            _patch_value(value[jdx], nv)
                                        else:
                                            value[jdx] = nv

                                    else:
                                        value.append(nv)
                        else:
                            value[idx] = v
                    else:
                        if isinstance(v, list):
                            value.extend(v)
                        else:
                            value.append(v)
        if remove:
            # To avoid index shift, we sort the keys in reverse order
            for k in sorted(remove.keys(), reverse=True):
                if isinstance(k, int) and 0 <= k < len(value):
                    v = remove[k]
                    if isinstance(v, dict):
                        _patch_value(value[k], remove=v)
                    else:
                        del value[k]
    return original_value

File: frontend/taipy-gui/src/context/patch.ts
/*
 * Copyright 2021-2025 Avaiga Private Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

export interface PatchChange {
    [key: string | number]: string | number | boolean | null | PatchChange;
}
export interface PatchRemove {
    [key: string | number]: null | PatchRemove;
}

const RE_NUMBER = /^-?\d+$/;

export const patchValue = <T>(toBePatched: T, change?: PatchChange, remove?: PatchRemove): T => {
    let patchedValue = toBePatched;
    if (change) {
        // Apply changes
        Object.entries(change).forEach(([k, v]) => {
            const sIdx = Number(k);
            const idx = sIdx < 0 ? -1 - sIdx : sIdx;
            const insert = sIdx < 0;
            if (
                Array.isArray(toBePatched) &&
                RE_NUMBER.test(k) &&
                (insert ? toBePatched.length > idx : toBePatched.length >= idx)
            ) {
                // TODO deal with _tp_index
                const oldValue = toBePatched[idx];
                if (oldValue !== v) {
                    if (Array.isArray(v) && v.length > 0) {
                        if (insert) {
                            if (idx >= (patchedValue as Array<unknown>).length) {
                                patchedValue = [...(patchedValue as Array<unknown>), ...v] as T;
                            } else {
                                if (patchedValue === toBePatched) {
                                    patchedValue = [...toBePatched] as T;
                                }
                                (patchedValue as Array<unknown>).splice(idx, 0, ...v);
                            }
                        } else {
                            if (patchedValue === toBePatched) {
                                patchedValue = [...toBePatched] as T;
                            }
                            const newArray = (v as Array<unknown>).map((item, j) => {
                                const oldItem = (toBePatched as Array<unknown>)[idx + j];
                                if (
                                    typeof item === "object" &&
                                    item !== null &&
                                    typeof oldItem === "object" &&
                                    oldItem !== null
                                ) {
                                    return patchValue(oldItem, item as PatchChange);
                                } else {
                                    return item;
                                }
                            });
                            (patchedValue as Array<unknown>).splice(idx, newArray.length, ...newArray);
                        }
                    } else if (
                        oldValue !== null &&
                        typeof oldValue === "object" &&
                        v !== null &&
                        typeof v === "object"
                    ) {
                        const newValue = patchValue(oldValue, v);
                        if (newValue !== oldValue) {
                            if (patchedValue === toBePatched) {
                                patchedValue = [...toBePatched] as T;
                            }
                            (patchedValue as Array<unknown>)[idx] = newValue;
                        }
                    } else if (
                        (oldValue === null || typeof oldValue !== "object") &&
                        (v === null || typeof v !== "object")
                    ) {
                        if (idx >= (patchedValue as Array<unknown>).length) {
                            patchedValue = [...(patchedValue as Array<unknown>), v] as T;
                        } else {
                            if (patchedValue === toBePatched) {
                                patchedValue = [...toBePatched] as T;
                            }
                            (patchedValue as Array<unknown>)[idx] = v;
                        }
                    }
                }
            } else if (toBePatched && typeof toBePatched === "object" && !Array.isArray(toBePatched)) {
                const oldValue = (toBePatched as Record<string, unknown>)[k];
                if (oldValue !== v) {
                    if (Array.isArray(v) && v.length > 0) {
                        if (patchedValue === toBePatched) {
                            patchedValue = { ...toBePatched };
                        }
                        (patchedValue as Record<string, unknown>)[k] = v;
                    } else if (
                        oldValue !== null &&
                        typeof oldValue === "object" &&
                        v !== null &&
                        typeof v === "object"
                    ) {
                        const newValue = patchValue(oldValue, v);
                        if (newValue !== oldValue) {
                            if (patchedValue === toBePatched) {
                                patchedValue = { ...toBePatched };
                            }
                            (patchedValue as Record<string, unknown>)[k] = newValue;
                        }
                    } else if (
                        (oldValue === null || typeof oldValue !== "object")
                    ) {
                        if (patchedValue === toBePatched) {
                            patchedValue = { ...toBePatched };
                        }
                        (patchedValue as Record<string, unknown>)[k] = v;
                    }
                }
            }
        });
    }
    if (remove) {
        // Apply removals
        Object.entries(remove).forEach(([k, v]) => {
            const idx = Number(k);
            if (RE_NUMBER.test(k) && Array.isArray(toBePatched) && toBePatched.length > idx) {
                const oldValue = (toBePatched as Array<unknown>)[idx];
                if (oldValue !== undefined) {
                    if (v !== null && typeof v === "object" && !Array.isArray(v)) {
                        const newValue = patchValue(oldValue, undefined, v);
                        if (newValue !== oldValue) {
                            if (patchedValue === toBePatched) {
                                patchedValue = [...toBePatched] as T;
                            }
                            (patchedValue as Array<unknown>)[idx] = newValue;
                        }
                    } else {
                        if (patchedValue === toBePatched) {
                            patchedValue = [...toBePatched] as T;
                        }
                        (patchedValue as Array<unknown>).splice(idx, 1);
                    }
                }
            } else if (toBePatched && typeof toBePatched === "object" && !Array.isArray(toBePatched)) {
                const oldValue = (toBePatched as Record<string, unknown>)[k];
                if (oldValue !== undefined) {
                    if (v !== null && typeof v === "object" && !Array.isArray(v)) {
                        const newValue = patchValue(oldValue, undefined, v);
                        if (newValue !== oldValue) {
                            if (patchedValue === toBePatched) {
                                patchedValue = { ...toBePatched };
                            }
                            (patchedValue as Record<string, unknown>)[k] = newValue;
                        }
                    } else {
                        if (patchedValue === toBePatched) {
                            patchedValue = { ...toBePatched };
                        }
                        delete (patchedValue as Record<string, unknown>)[k];
                    }
                }
            }
        });
    }
    return patchedValue;
};
